stock_reserve_rule/models/stock_reserve_rule.py

```diff
@@ -0,0 +1,278 @@
+# Copyright 2019 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+from odoo import fields, models
+from odoo.osv import expression
+from odoo.tools.float_utils import float_compare
+from odoo.tools.safe_eval import safe_eval
+
+
+def _default_sequence(record):
+    maxrule = record.search([], order="sequence desc", limit=1)
+    if maxrule:
+        return maxrule.sequence + 10
+    else:
+        return 0
+
+
+class StockReserveRule(models.Model):
+    """Rules for stock reservations
+
+    Each rule can have many removal rules, they configure the conditions and
+    advanced removal strategies to apply on a specific location (sub-location
+    of the rule).
+
+    The rules are selected for a move based on their source location and a
+    configurable domain on the rule.
+    """
+
+    _name = "stock.reserve.rule"
+    _description = "Stock Reservation Rule"
+    _order = "sequence, id"
+
+    name = fields.Char(string="Description", required=True)
+    sequence = fields.Integer(default=lambda s: _default_sequence(s))
+    active = fields.Boolean(default=True)
+    company_id = fields.Many2one(
+        comodel_name="res.company", default=lambda self: self.env.user.company_id.id
+    )
+
+    location_id = fields.Many2one(comodel_name="stock.location", required=True)
+    fallback_location_id = fields.Many2one(
+        comodel_name="stock.location",
+        help="If all removal rules are exhausted, try to reserve in this "
+        "location. When empty, the fallback happens in any of the move's "
+        "source sub-locations.",
+    )
+
+    rule_removal_ids = fields.One2many(
+        comodel_name="stock.reserve.rule.removal", inverse_name="rule_id"
+    )
+
+    rule_domain = fields.Char(
+        string="Rule Domain",
+        default=[],
+        help="Domain based on Stock Moves, to define if the "
+        "rule is applicable or not.",
+    )
+
+    def _rules_for_location(self, location):
+        return self.search([("location_id", "parent_of", location.id)])
```

Comment:
I think a move should never look for a rule that could reserve goods in a location that is not a child of the move source location.